features/dnsresolver/netty/src/test/java/org/opennms/netmgt/dnsresolver/netty/NettyDnsResolverTest.java

```diff
@@ -0,0 +1,194 @@
+            assertThat(e.getCause(), is(instanceOf(DnsNameResolverTimeoutException.class)));
+        }
+    }
+
+    @Test
+    public void canTriggerOpenCircuit() throws InterruptedException, TimeoutException {
+        // Reinitialize the resolver using a non-routable address as the target - we want the queries to fail due to timeouts
+        dnsResolver.destroy();
+        dnsResolver.setNameservers(InetAddressUtils.str(InetAddressUtils.UNPINGABLE_ADDRESS));
+        dnsResolver.init();
+
+        // Now trigger enough requests to open the circuit breaker
+        final int N = 2 * dnsResolver.getCircuitBreaker().getCircuitBreakerConfig().getRingBufferSizeInClosedState();
+        final CompletableFuture futures[] = new CompletableFuture[N];
+        for (int i = 0; i < N; i++) {
+            futures[i] = dnsResolver.reverseLookup(InetAddressUtils.addr("fe80::"));
+        }
+
+        // Wait for the requests to complete
+        try {
+            CompletableFuture.allOf(futures)
+                    // This should not take longer than the query timeout
+                    .get(2 * dnsResolver.getQueryTimeoutMillis(), TimeUnit.MILLISECONDS);
+            fail("Expected an ExecutionException to be thrown");
+        } catch (ExecutionException e) {
+            // pass
+        }
+
+        // The circuit breaker should be in a open state now and start rejecting calls
+        try {
+            dnsResolver.reverseLookup(InetAddressUtils.addr("fe80::")).get();
+            fail("Expected an CallNotPermittedException to be thrown");
+        } catch (ExecutionException e) {
+            assertThat(e.getCause(), is(instanceOf(CallNotPermittedException.class)));
+        }
+    }
+
+    @Test
+    @Ignore
+    public void reverseLookupUsingPublicDnsServers() throws UnknownHostException, ExecutionException, InterruptedException {
+        // Use the system's servers
+        dnsResolver.destroy();
+        dnsResolver.setNameservers("4.2.2.2");
+        dnsResolver.init();
+
+        final Optional<String> hostname1 = dnsResolver.reverseLookup(InetAddress.getByAddress(new byte[]{1, 1, 1, 1})).get();
+        assertEquals("one.one.one.one", hostname1.get());
+
+        final Optional<String> hostname2 = dnsResolver.reverseLookup(InetAddressUtils.addr("1.1.1.1")).get();
+        assertEquals("one.one.one.one", hostname2.get());
+
+        final Optional<String> hostname3 = dnsResolver.reverseLookup(InetAddressUtils.addr("2606:4700:4700::1111")).get();
+        assertEquals("one.one.one.one", hostname3.get());
+
+        // 198.51.100.0/24 should be TEST-NET-2 (see RFC #5737). Should fail...
+        final Optional<String> hostname4 = dnsResolver.reverseLookup(InetAddressUtils.addr("198.51.100.1")).get();
+        assertEquals(Optional.empty(), hostname4);
+
+        final Optional<String> hostname5 = dnsResolver.reverseLookup(InetAddressUtils.addr("fe80::")).get();
+        assertEquals(Optional.empty(), hostname5);
+    }
+
```

Comment:
Maybe add a test to check that a comma-separated list is parsed correctly. I think spaces are currently a problem for IP and port, e.g. "8.8.8.8:53 ,1.1.1.1, 1.1.2.2"

Reply:
Test added.